lib/urireader.cpp

```diff
@@ -3,11 +3,63 @@
 #include "timing.h"
 #include "urireader.h"
 #include "util.h"
+#include "encode.h"
 #include <sys/mman.h>
 #include <sys/stat.h>
+#include <cstdlib>
+#include <ctime>
 
 namespace HTTP{
 
+  // Helper for Date header
+  std::string dateToString() {
+    char buffer[80];
+    time_t rawtime;
+    struct tm * timeinfo;
+    time(&rawtime);
+    timeinfo = localtime(&rawtime);
+    strftime(buffer, sizeof(buffer), "%a, %d %b %Y %H:%M:%S %z", timeinfo);
+    return std::string(buffer);
+  }
+
+  // Input url == s3+https://s3_key:secret@storage.googleapis.com/alexk-dms-upload-test/testvideo.ts
+  // Transform to: 
+  // url=https://storage.googleapis.com/alexk-dms-upload-test/testvideo.ts
+  // header Date: ${dateToString()}
+  // header Authorization: AWS ${url.user}:${signature}
+  bool transformS3ToHttp(HTTP::Downloader& downloader, HTTP::URL& url) {
+    // Adjust URL
+    url.protocol = url.protocol.erase(0, 3); // remove "s3+" prefix
+    std::string accessKey(url.user), secret(url.pass), date(dateToString());
+    url.user = "";
+    url.pass = "";
+    std::string requestPath = "/" + Encodings::URL::encode(url.path, "/:=@[]#?&");
+    if(url.args.size()) requestPath += "?" + url.args;
+    // Calculate Authorization data
+    std::string toSign = "GET\n\n\n" + date + "\n" + requestPath;
+    unsigned char signature[MBEDTLS_MD_MAX_SIZE];
+    const int sha1Size = 20;
+    mbedtls_md_context_t md_ctx = {0};
+    // TODO: When we use MBEDTLS_MD_SHA512 ? Consult documentation/code
+    const mbedtls_md_info_t *md_info = mbedtls_md_info_from_type(MBEDTLS_MD_SHA1);
```

Comment:
When compiling without mbedtls (`NOSSL` will be defined), this code should not be compiled in either. (Yes, I know that currently Mist doesn't compile with `NOSSL` set, but no need to make that problem even worse, right? 😇)